File: webhooks/python/samples/room.restored.py
# Room Restore - Notify Members, Admin, Audit Access

from flask import Flask, request, abort
import hmac
import hashlib
import requests
import os
from datetime import datetime

app = Flask(__name__)

SENSITIVE_ROOM_KEYWORDS = ["hr", "finance", "legal", "clients", "external"]

# Generate HMAC-SHA256 signature
def get_secret_hash(secret, payload):
    return "sha256=" + hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()

@app.route("/webhook", methods=["POST"])
def webhook():
    signature = request.headers.get("x-docspace-signature-256")
    computed_signature = get_secret_hash(
        os.environ.get("WEBHOOK_SECRET_KEY"),
        request.get_data()
    )

    if signature != computed_signature:
        app.logger.warning("Invalid signature")
        abort(401)

    event = request.json.get("event", {})
    payload = request.json.get("payload", {})

    trigger = event.get("trigger")
    if trigger != "room.restored":
        return "", 200

    room_title = payload.get("title")
    room_id = payload.get("rootId")
    created_by = payload.get("createBy")
    lower_title = room_title.lower() if room_title else ""
    timestamp = datetime.utcnow().isoformat()

    app.logger.info(f'Room restored: "{room_title}" ({room_id})')

    try:
        # Step 1: Notify all room members
        share_res = requests.get(
            f"{os.environ.get('PORTAL_URL')}/api/2.0/files/rooms/{room_id}/share",
            headers={"Authorization": f"Bearer {os.environ.get('API_KEY')}"}
        ).json()

        users = share_res.get("access", [])

        for user in users:
            email = user.get("sharedTo", {}).get("email")
            if not email:
                continue

            requests.post(
                os.environ.get("NOTIFY_WEBHOOK_URL"),
                json={
                    "to": email,
                    "subject": "Room restored",
                    "message": f'The room "{room_title}" has been restored and is now available.'
                },
                headers={"Content-Type": "application/json"}
            )

            app.logger.info(f"User notified: {email}")

        # Step 2: If sensitive room - notify admin with access list
        is_sensitive = any(keyword in lower_title for keyword in SENSITIVE_ROOM_KEYWORDS)
        if is_sensitive:
            restorer_info = requests.get(
                f"{os.environ.get('PORTAL_URL')}/api/2.0/people/{created_by}",
                headers={"Authorization": f"Bearer {os.environ.get('API_KEY')}"}
            ).json()

            restorer_name = f"{restorer_info.get('firstName', '')} {restorer_info.get('lastName', '')}".strip()
            restorer_email = restorer_info.get("email", "[unknown]")

            access_list = "\n".join([
                f"- {user.get('sharedTo', {}).get('email', '[no email]')}"
                + (" (external)" if not user.get('sharedTo', {}).get('email', '').endswith("@onlyoffice.com") else "")
                for user in users
            ])

            message = (
                f"Sensitive room \"{room_title}\" has been restored.\n\n"
                f"Restored by: {restorer_name} ({restorer_email})\n"
                f"Room ID: {room_id}\n"
                f"Date: {timestamp}\n\n"
                f"Room Access List:\n{access_list}"
            )

            requests.post(
                os.environ.get("NOTIFY_WEBHOOK_URL"),
                json={
                    "to": os.environ.get("ADMIN_EMAIL"),
                    "subject": "Sensitive room restored",
                    "message": message
                },
                headers={"Content-Type": "application/json"}
            )

            app.logger.info("Admin notified with access list")

        return "", 200

    except Exception as err:
        app.logger.error(f"Error in room.restored handler: {err}")
        return "", 500

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 3000))
    app.run(host="0.0.0.0", port=port)

File: webhooks/nodejs/samples/room.restored.js
// Room Restore - Notify Members, Admin, Audit Access

const express = require("express");
const crypto = require("crypto");
const axios = require("axios");
require("dotenv").config();

const app = express();
app.use(express.json({ verify: rawBodySaver }));

function rawBodySaver(req, res, buf) {
  req.rawBody = buf.toString();
}

function getSecretHash(secret, payload) {
  return (
    "sha256=" +
    crypto.createHmac("sha256", secret).update(payload).digest("hex")
  );
}

const SENSITIVE_ROOM_KEYWORDS = ["hr", "finance", "legal", "clients", "external"];

app.post("/webhook", async (req, res) => {
  const signature = req.headers["x-docspace-signature-256"];
  const computedSignature = getSecretHash(
    process.env.WEBHOOK_SECRET_KEY,
    req.rawBody
  );
  if (signature !== computedSignature) return res.sendStatus(401);

  const { trigger } = req.body.event;
  if (trigger !== "room.restored") return res.sendStatus(200);

  const { title: roomTitle, rootId: roomId, createBy } = req.body.payload;
  const lowerTitle = roomTitle.toLowerCase();
  const timestamp = new Date().toISOString();

  console.log(`Room restored: "${roomTitle}" (${roomId})`);

  try {
    // Step 1: Notify all room members
    const shareRes = await axios.get(
      `${process.env.PORTAL_URL}/api/2.0/files/rooms/${roomId}/share`,
      {
        headers: {
          Authorization: `Bearer ${process.env.API_KEY}`
        }
      }
    );

    const users = shareRes.data.access || [];

    for (const user of users) {
      const email = user.sharedTo?.email;
      if (!email) continue;

      await axios.post(
        process.env.NOTIFY_WEBHOOK_URL,
        {
          to: email,
          subject: "Room restored",
          message: `The room "${roomTitle}" has been restored and is now available.`
        },
        {
          headers: { "Content-Type": "application/json" }
        }
      );

      console.log(`User notified: ${email}`);
    }

    // Step 2: If sensitive room - notify admin with access list
    const isSensitive = SENSITIVE_ROOM_KEYWORDS.some(kw => lowerTitle.includes(kw));
    if (isSensitive) {
      // Get restorer info
      const restorerInfo = await axios.get(
        `${process.env.PORTAL_URL}/api/2.0/people/${createBy}`,
        {
          headers: { Authorization: `Bearer ${process.env.API_KEY}` }
        }
      );

      const restorerName = `${restorerInfo.data?.firstName || ""} ${restorerInfo.data?.lastName || ""}`.trim();
      const restorerEmail = restorerInfo.data?.email || "[unknown]";

      // Build access list with external user flag
      const accessList = users.map(user => {
        const email = user.sharedTo?.email || "[no email]";
        const isExternal = !email.endsWith("@onlyoffice.com");
        return `- ${email}${isExternal ? " (external)" : ""}`;
      }).join("\n");

      const message = `
Sensitive room "${roomTitle}" has been restored.

Restored by: ${restorerName} (${restorerEmail})
Room ID: ${roomId}
Date: ${timestamp}

Room Access List:
${accessList}
      `.trim();

      await axios.post(
        process.env.NOTIFY_WEBHOOK_URL,
        {
          to: process.env.ADMIN_EMAIL,
          subject: "Sensitive room restored",
          message
        },
        {
          headers: { "Content-Type": "application/json" }
        }
      );

      console.log("Admin notified with access list");
    }

    return res.sendStatus(200);
  } catch (err) {
    console.error("Error in room.restored handler:", err.response?.data || err.message);
    return res.sendStatus(500);
  }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Webhook server listening on port ${PORT}`);
});
